File: backend/app/schemas/credit.py
"""Credit system Pydantic schemas."""
from datetime import datetime
from decimal import Decimal
from typing import Optional, Any

from pydantic import BaseModel, Field


class CreditBalanceResponse(BaseModel):
    """Response schema for credit balance."""

    user_id: int
    balance: Decimal = Field(description="Current credit balance")
    lifetime_purchased: Decimal = Field(description="Total credits purchased all-time")
    lifetime_spent: Decimal = Field(description="Total credits spent all-time")
    last_purchase_date: Optional[datetime] = Field(None, description="Last purchase timestamp")
    last_transaction_date: Optional[datetime] = Field(None, description="Last transaction timestamp")

    class Config:
        from_attributes = True


class PurchaseRequest(BaseModel):
    """Request schema for credit purchase."""

    package: str = Field(..., description="Package type: starter, basic, standard, premium")
    payment_method: str = Field(..., description="Payment method: bank_transfer, qr_code")
    return_url: Optional[str] = Field(None, description="URL to redirect after payment")


class BankAccountInfo(BaseModel):
    """Bank account information for payment."""

    bank_name: str
    account_number: str
    account_name: str
    transfer_content: str


class PurchaseResponse(BaseModel):
    """Response schema for credit purchase."""

    purchase_id: str
    package: str
    amount_vnd: int
    credits: Decimal
    payment_url: str
    qr_code: str = Field(description="Base64 encoded QR code image")
    bank_account: BankAccountInfo
    expires_at: datetime
    status: str


class CreditTransactionItem(BaseModel):
    """Single credit transaction item."""

    id: str
    type: str = Field(description="Transaction type: purchase, usage, refund, adjustment")
    amount: Decimal = Field(description="Amount (positive for credits, negative for debits)")
    balance_after: Decimal = Field(description="Balance after this transaction")
    description: str
    reference_id: Optional[str] = None
    extra_data: Optional[dict[str, Any]] = None
    created_at: datetime

    class Config:
        from_attributes = True


class TransactionHistoryResponse(BaseModel):
    """Response schema for transaction history."""

    total: int = Field(description="Total number of transactions matching filter")
    page: int = Field(description="Current page number")
    per_page: int = Field(description="Items per page")
    pages: int = Field(description="Total number of pages")
    transactions: list[CreditTransactionItem]

File: frontend/src/types/credit.ts
/**
 * Credit system TypeScript types
 */

export type PackageType = 'starter' | 'basic' | 'standard' | 'premium'
export type PaymentMethod = 'bank_transfer' | 'qr_code'
export type TransactionType = 'purchase' | 'usage' | 'refund' | 'adjustment'

export interface CreditBalance {
  user_id: number
  balance: number
  lifetime_purchased: number
  lifetime_spent: number
  last_purchase_date: string | null
  last_transaction_date: string | null
}

export interface PurchaseRequest {
  package: PackageType
  payment_method: PaymentMethod
  return_url?: string
}

export interface BankAccountInfo {
  bank_name: string
  account_number: string
  account_name: string
  transfer_content: string
}

export interface PurchaseResponse {
  purchase_id: string
  package: PackageType
  amount_vnd: number
  credits: number
  payment_url: string
  qr_code: string
  bank_account: BankAccountInfo
  expires_at: string
  status: string
}

export interface CreditTransaction {
  id: string
  type: TransactionType
  amount: number
  balance_after: number
  description: string
  reference_id: string | null
  extra_data: Record<string, any> | null
  created_at: string
}

export interface TransactionHistory {
  total: number
  page: number
  per_page: number
  pages: number
  transactions: CreditTransaction[]
}
